### PR TITLE
[NEW] Api provider

### LINKED ISSUE #12: Implement the Webhooks
Allow endpoints for Rocket.Chat Apps which will be fully functioning webhooks which support get, post, put, delete, etc.

TypeScript Definition: https://github.com/RocketChat/Rocket.Chat.Apps-ts-definition/tree/master/src/webhooks

### PR DESCRIPTION
**Pull request by rodrigok**
Closes #12

Rocket.Chat bridge implementation https://github.com/RocketChat/Rocket.Chat/pull/11938

TODO
- [x] List provided webhooks on app's details page to be easy to copy
- [x] Add property to define hook as private or public
- [x] Add property to define hook security level
- [x] Property to provide curl samples for each hook

IMPROVEMENTS
- [ ] Implement security level TOKEN_AUTH
- [ ] Implement security level SHARED_SECRET

Webhook admin UI
image

Implementation example
```typescript
// index.ts
import {
    IConfigurationExtend,
    ILogger,
} from '@rocket.chat/apps-engine/definition/accessors';
import { ApiSecurity, ApiVisibility } from '@rocket.chat/apps-engine/definition/api/IApi';
import { App } from '@rocket.chat/apps-engine/definition/App';
import { IAppInfo } from '@rocket.chat/apps-engine/definition/metadata';

import { MessageEndpoint } from './endpoints';

export class JitsiApp extends App {
    constructor(info: IAppInfo, logger: ILogger) {
        super(info, logger);
    }

    protected async extendConfiguration(configuration: IConfigurationExtend): Promise<void> {
        await configuration.api.provideApi({
            visibility: ApiVisibility.PRIVATE,
            security: ApiSecurity.UNSECURE,
            endpoints: [
                new MessageEndpoint(),
            ],
        });
    }
}
```
```typescript
// endpoints.ts
import { HttpStatusCode, IHttp, IModify, IPersistence, IRead } from '@rocket.chat/apps-engine/definition/accessors';
import { IApiRequest, IApiResponse } from '@rocket.chat/apps-engine/definition/api';
import { IApiEndpoint } from '@rocket.chat/apps-engine/definition/api/IApi';
import { example } from '@rocket.chat/apps-engine/definition/api/IApiExample';
import { IApiEndpointInfo } from '@rocket.chat/apps-engine/definition/api/IApiEndpointInfo';

export class MessageEndpoint implements IApiEndpoint {
    public path: string = 'message/:id';

    @example({
        params: {
            id: 'message_id',
        },
        query: {
            asd: '123',
        },
        headers: {
            'X-Auth': 'asd: 123',
        },
    })
    public async get(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse> {
        return {
            status: HttpStatusCode.OK,
            headers: {
                'X-Auth': 'asd: 123',
            },
            content: {
                test: request.query.asd,
            },
        };
    }

    @example({
        params: {
            id: 'message_id',
        },
        query: {
            asd: '123',
        },
        content: {
            content: 1,
        },
    })
    public async post(request: IApiRequest, endpoint: IApiEndpointInfo, read: IRead, modify: IModify, http: IHttp, persis: IPersistence): Promise<IApiResponse> {
        return {
            status: HttpStatusCode.OK,
            content: request.content,
        };
    }
}
```

### CODE REVIEW

**Comment by rodrigok** [comment]
@graywolf336 can you provide some examples?

**Comment by piotrkochan** [comment]
awesome!

File: src/server/managers/AppWebhook.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    'get',
+    'post',
+    'put',
+    'delete',
+    'head',
+    'options',
+    'patch',
+];
+
+export class AppWebhook {
```
Comment (graywolf336) [inline src/server/managers/AppWebhook.ts]: What's the purpose of having two `AppWebhook` items within the managers folder?

Comment (geekgonecrazy) [reply src/server/managers/AppWebhook.ts]: @rodrigok @graywolf336 renamed from webhook to api.

**Comment by rodrigok** [comment]
@geekgonecrazy awesome, can you remove the docs changes? I'm planning to move docs to a dedicated branch to prevent this mess on PRs as Bradley suggested some time ago.

**Comment by geekgonecrazy** [comment]
@rodrigok done :)

File: package.json [hunk 2]
```diff
@@ -57,19 +56,20 @@
...
     "pre-commit": "^1.2.2",
     "tap-bark": "^1.0.0",
-    "ts-node": "^6.0.2",
-    "tslint": "^5.9.1",
+    "ts-node": "^6.2.0",
+    "tslint": "^5.11.0",
     "typedoc": "^0.11.1"
   },
   "dependencies": {
+    "@types/node": "^10.9.4",
```
Comment (graywolf336) [inline package.json]: Any reason the this has been added to the dependencies?

Comment (rodrigok) [reply package.json]: Yes, was needed to allow the internal module imports, but I found another solution for that. So I'll remove it.

File: src/definition/api/IApi.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Provide enpoints for this api registry
+     */
+    endpoints: Array<IApiEndpoint>;
+}
+
+export enum ApiVisibility {
+    /**
+     * Public api's have a fixed URL format, they are easy to remember
+     * but they are easy to be hacked as well.
```
Comment (graywolf336) [inline src/definition/api/IApi.ts]: Can we change the `hacked` to `guessed`? Maybe reword this to be:

`A public Api has a fixed format for a url. Using it enables an easy to remember structure, however, it also means the url is intelligently guessed. As a result, we recommend having some sort of security setup if you must have a public api. Whether you use the provided security, ApiSecurity, or implement your own.`

File: src/definition/api/IApi.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { IApiEndpointInfo } from './IEndpoint';
+import { IApiRequest } from './IRequest';
+import { IApiResponse } from './IResponse';
+
+export interface IApiEndpoint {
+    /**
+     * Path to complete the api URL. Example: https://{your-server-address}/api/apps/public/{your-app-id}/{path}
+     * or https://{your-server-address}/api/apps/private/{your-app-id}/{private-hash}/{path}
+     */
+    path: string;
```
Comment (graywolf336) [inline src/definition/api/IApi.ts]: What would you think about allowing param style route definitions? For example: `/contacts/:contactId/numbers` would result in the passed in `contactId` being a param. Maybe this can be a future plan for improvements

File: src/definition/api/IApi.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import { IHttp, IModify, IPersistence, IRead } from '../accessors';
+import { IApiExample } from './IApiExample';
+import { IApiEndpointInfo } from './IEndpoint';
+import { IApiRequest } from './IRequest';
+import { IApiResponse } from './IResponse';
+
+export interface IApiEndpoint {
```
Comment (graywolf336) [inline src/definition/api/IApi.ts]: I am still of the opinion this should be a class. That way we can have helper methods for converting various things from one to another. For example, the `IApiRequest` has `content: any`. But let's be real, most people are going to expect that to be a json value with a format of `{ [key: string]: string }`. So, we could have helper methods for the TypeScript to assume that something is what they expect it to be. Also, for converting from a string to json. Could also have helper methods for common things such as a successful response `return this.success()` or `return this.json(statusCode: number, data: object)`.

File: src/server/managers/AppApiManager.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * @param appId the app which is providing the api
+     * @param path the path to be executed in app's api's
+     * @param request the request data to be evaluated byt the app
+     */
+    public executeApi(appId: string, path: string, request: IApiRequest): Promise<IApiResponse> {
+        const api = this.providedApis.get(appId).get(path);
+
+        if (!api) {
+            return;
```
Comment (graywolf336) [inline src/server/managers/AppApiManager.ts]: I think we should be returning a response here, even if it's just a 400 bad request. That way the bridged system doesn't have to handle nothing being returned and they actually get back a response when they call this method.

File: src/server/managers/AppApiManager.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        const app = this.manager.getOneById(appId);
+
+        if (!app || AppStatusUtils.isDisabled(app.getStatus())) {
+            // Just in case someone decides to do something they shouldn't
+            // let's ensure the app actually exists
+            return;
```
Comment (graywolf336) [inline src/server/managers/AppApiManager.ts]: Same here as my comment above

File: src/definition/api/IApi.ts (hunk 4 above)
Comment (rodrigok) [reply src/definition/api/IApi.ts]: This already works 😄

File: src/definition/accessors/IApiExtend.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * This accessor provides methods for adding a custom api.
+ * It is provided during the initialization of your App
+ */
+
+export interface IApiExtend {
+    /**
+     * Adds an api which can be called by external services lateron.
+     * Should an api already exists an error will be thrown.
```
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/accessors/IApiExtend.ts]: Idk if I understood your point, but you are able to provide more than one API, mainly if you need different properties like visibility and security. Is that what you were meaning?

File: src/definition/api/IApi.ts (hunk 5 above)
Comment (rodrigok) [reply src/definition/api/IApi.ts]: I see your points, would it be better to have a class in place of the interface or a class that implements the interface named `ApiEndpoint`?

File: src/definition/api/IEndpoint.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export interface IApiEndpointInfo {
```
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/api/IEndpoint.ts]: 👍

File: src/definition/api/IApi.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+     * No security check will be executed agains the calls made to this URL
+     */
+    UNSECURE,
+    /**
+     * Only calls containing a valid token will be able to execute the api
+     * Mutiple tokens can be generated to access the api, by default one
+     * will be generated automatically.
+     * @param `X-Auth-Token`
+     */
+    // AUTH_TOKEN,
```
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/api/IApi.ts]: Well, the AUTH_TOKEN is already documented there, the idea is to have options to handle authentication automatically for the apps. But it's not clear the options for me now.

File: src/definition/api/IApi.ts (hunk 5 above)
Comment (graywolf336) [reply src/definition/api/IApi.ts]: Now that's a good question, probably doesn't matter too much. Thinking about it though, might be best to just have one so there is less source for confusion when an external developer tries to look into the code to see how to do things.

File: src/server/managers/AppApi.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    public readonly appId: string;
+    public readonly hash?: string;
+    public readonly implementedMethods: Array<string>;
+
+    constructor(public app: ProxiedApp, public api: IApi, public endpoint: IApiEndpoint) {
+        this.appId = app.getID();
+
+        switch (this.api.visibility) {
+            case ApiVisibility.PUBLIC:
+                this.basePath = `/api/apps/public/${app.getID()}`;
```
Comment (graywolf336) [inline src/server/managers/AppApi.ts]: @rodrigok what would you think about us making the `nameSlug` of the Apps to be unique and use it instead of the App's ID? This would help with the transition over to #56 (package.json) to slowly get rid of the usage of the id. As a result, and where the idea came to me from, the "endpoints" and urls generated/provided will be easier to use (and shorter) instead of having a long id in the url.

File: src/definition/api/IApi.ts (hunk 4 above)
Comment (graywolf336) [reply src/definition/api/IApi.ts]: @rodrigok sweet! Maybe you can add some documentation so other developers are aware of it and can make usage of it 😃

File: src/definition/api/IApi.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (graywolf336) [reply src/definition/api/IApi.ts]: It's commented out, so it won't show up in the generated docs if I'm not mistaken..yeah I'm not sure the best route to go yet without putting forth much effort in thoughts.

File: src/definition/accessors/IApiExtend.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/accessors/IApiExtend.ts]: By the API we mean an endpoint, you can configure more than one API but containing different endpoints

File: src/definition/api/IApi.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/api/IApi.ts]: I don't want to have it in docs since we don't have it implemented yet, it's basically a reminder that we will have more options there.

File: src/server/managers/AppApi.ts (hunk 11 above)
Comment (rodrigok) [reply src/server/managers/AppApi.ts]: I wan't to move the ID to be a unique ID based on name as NPM does, that way we keep using the ID here, just change how the ID would be generated.

File: src/definition/accessors/IApiExtend.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (graywolf336) [reply src/definition/accessors/IApiExtend.ts]: So, you can have more than one API but different endpoints? If I call this method more than once, each time containing two different endpoints, will it error out or will it succeed?

File: src/server/managers/AppApi.ts (hunk 11 above)
Comment (graywolf336) [reply src/server/managers/AppApi.ts]: Right, so if we go with the name slug or a special smaller ID such as `author:app-name` then we could generate shorter urls and that would be a unique ID since the author couldn't have two of the same app names but two people could have the same App name as their author name would be different.

File: src/definition/accessors/IApiExtend.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rodrigok) [reply src/definition/accessors/IApiExtend.ts]: Will succeed

File: src/server/managers/AppApi.ts (hunk 11 above)
Comment (rodrigok) [reply src/server/managers/AppApi.ts]: Yes, but that will be in the future

**Comment by meomay503** [comment]
HI @graywolf336 @rodrigok, does your team have a plan to implement the other security level for the API in the next releases?

**Comment by rodrigok** [comment]
@meomay503, unfortunately, no, it's not in our short roadmap.
